projects/graph/graph.py

```diff
@@ -16,42 +15,66 @@ def add_vertex(self, vertex):
         """
         Add a vertex to the graph.
         """
-        pass  # TODO
+        self.vertices[vertex] = set()
     def add_edge(self, v1, v2):
         """
         Add a directed edge to the graph.
         """
-        if v1 not self.vertices:
+        if v1 not in self.vertices:
             self.add_vertex(v1);
+
         if v2 not in self.vertices:
             self.add_vertex(v2);
-        self.vertices[v1].add(v2);
+
+        self.vertices[v1].add(v2)
         
         
     def bft(self, starting_vertex):
         """
         Print each vertex in breadth-first order
         beginning from starting_vertex.
         """
-       #make a queue
-       queue = Queue()
-       #make a visited set
-       visited = set()
-       # put starting vertex in he queue
-       queue.enqueue(starting_vertex)
-       #while q isn\t empty
-       while queue.size():
-       # dequeue the item, it is our current item
+        #make a queue
+        queue = Queue()
+        #make a visited set
+        visited = set()
+        # put starting vertex in he queue
+        queue.enqueue(starting_vertex)
+        #while q isn\t empty
+        while queue.size():
+        # dequeue the item, it is our current item
             node = queue.dequeue()
             print(node)
-       # mark current as visited
-            visited.add
+        # mark current as visited
+            visited.add(node)
+        # for each of the dequeued item's edge
+            for edge in self.vertices[node]:
```

Comment:
You should add some conditioning below this for loop to prevent vertices that had already been visited from being added to the queue again. If the graph is a bidirectional graph, this BFT solution code may execute the main while loop an infinite number of times.